src/Storages/MergeTree/MergeTreeSource.cpp

```diff
@@ -0,0 +1,220 @@
+        std::exception_ptr exception;
+
+        void finish()
+        {
+            stage = Stage::IsFinished;
+            event.write();
+        }
+
+        ChunkAndProgress getResult()
+        {
+            chassert(stage == Stage::IsFinished);
+            event.read();
+            stage = Stage::NotStarted;
+
+            if (exception)
+                std::rethrow_exception(exception);
+
+            return std::move(chunk);
+        }
+
+        friend struct AsyncReadingState;
+    };
+
+    std::shared_ptr<Control> start()
+    {
+        chassert(control->stage == Stage::NotStarted);
+        control->stage = Stage::InProgress;
+        return control;
+    }
+
+    void schedule(ThreadPool::Job job)
+    {
+        callback_runner(std::move(job), 0);
+    }
+
+    ChunkAndProgress getResult()
+    {
+        return control->getResult();
+    }
+
+    Stage getStage() const { return control->stage; }
+    int getFD() const { return control->event.fd; }
+
+    AsyncReadingState()
+    {
+        control = std::make_shared<Control>();
+        callback_runner = threadPoolCallbackRunner<void>(IOThreadPool::get(), "MergeTreeRead");
+    }
+
+    ~AsyncReadingState()
+    {
+        /// Here we wait for async task if needed.
+        /// ~AsyncReadingState and Control::finish can be run concurrently.
+        /// It's important to store std::shared_ptr<Control> into bg pool task.
+        /// Otherwise following is possible:
+        ///
+        ///  (executing thread)                         (bg pool thread)
+        ///                                             Control::finish()
+        ///                                             stage = Stage::IsFinished;
+        ///  ~MergeTreeBaseSelectProcessor()
+        ///  ~AsyncReadingState()
+        ///  control->stage != Stage::InProgress
+        ///  ~EventFD()
+        ///                                             event.write()
+        if (control->stage == Stage::InProgress)
+            control->event.read();
+    }
+
+private:
+    ThreadPoolCallbackRunner<void> callback_runner;
+    std::shared_ptr<Control> control;
+};
+#endif
+
+ISource::Status MergeTreeSource::prepare()
+{
+#if defined(OS_LINUX)
+    if (!async_reading_state)
+        return ISource::prepare();
+
+    /// Check if query was cancelled before returning Async status. Otherwise it may lead to infinite loop.
+    if (isCancelled())
+    {
+        getPort().finish();
+        return ISource::Status::Finished;
+    }
+
+    if (async_reading_state && async_reading_state->getStage() == AsyncReadingState::Stage::InProgress)
+        return ISource::Status::Async;
+#endif
+
+    return ISource::prepare();
+}
+
+
+std::optional<Chunk> MergeTreeSource::reportProgress(ChunkAndProgress chunk)
+{
+    if (chunk.num_read_rows || chunk.num_read_bytes)
+        progress(chunk.num_read_rows, chunk.num_read_bytes);
+
+    if (chunk.chunk.hasRows())
+        return std::move(chunk.chunk);
+
+    return {};
```

Comment:
```suggestion
    return chunk.chunk.hasRows() ? std::move(chunk.chunk) : Chunk{};
```

Reply:
btw it's not the same, cause Chunk() != std::nullopt.
And I prefer not using operator `?` :D

Reply:
yep, I overlooked `optional`.